### ISSUE TITLE
test dacapo-xalan_0 fail on JDK8 Arm32

### ISSUE THREAD

**Issue by Haroon-Khel** [adoptium/aqa-tests#3122]
test dacapo-xalan_0 in the jdk8 hs extended perf test suite fails on arm32

Latest failure https://ci.adoptopenjdk.net/job/Test_openjdk8_hs_extended.perf_arm_linux/26/tapResults/
I have rerun the test on other arm32 machines, still fails

```
===============================================
Running test dacapo-xalan_0 ...
===============================================
dacapo-xalan_0 Start Time: Wed Nov  3 13:19:34 2021 Epoch Time (ms): 1635945574285
variation: NoOptions
JVM_OPTIONS:  
Using scaled threading model. 32 processors detected, 32 threads used to drive the workload, in a possible range of [1,100]
===== DaCapo 9.12-MR1 xalan starting =====
; SystemID: file:///home/jenkins/workspace/Grinder/aqa-tests/TKG/output_16359455736097/dacapo-xalan_0/./scratch/xalan/xmlspec.xsl; Line#: 639; Column#: 29
javax.xml.transform.TransformerException: java.lang.RuntimeException: Programmer assertion is incorrect! - Can not setRoot on a non-iterated NodeSequence!
	at org.apache.xalan.transformer.TransformerImpl.executeChildTemplates(TransformerImpl.java:2405)
	at org.apache.xalan.templates.ElemLiteralResult.execute(ElemLiteralResult.java:1376)
	at org.apache.xalan.templates.ElemApplyTemplates.transformSelectedNodes(ElemApplyTemplates.java:395)
	at org.apache.xalan.templates.ElemApplyTemplates.execute(ElemApplyTemplates.java:178)
```

**Comment by sxa** [comment]
Seems less frequent on the odroid systems:

Machine | Result
---|---
test-docker-ubuntu2004-armv7l-1 | [FAILED 5/5](https://ci.adoptium.net/job/Grinder/7972/tapTestReport/)
test-docker-ubuntu2004-armv7l-2 | [FAILED 5/5](https://ci.adoptium.net/job/Grinder/7973/tapTestReport/)
test-docker-ubuntu2204-armv7l-1 | [FAILED 3/5](https://ci.adoptium.net/job/Grinder/7974/tapTestReport/)
test-sxa-armv7l-ubuntu2004-odroid-2 | [PASSED 5/5](https://ci.adoptium.net/job/Grinder/7976/tapTestReport/)
test-sxa-armv7l-ubuntu2004-odroid-1 | [TBC](https://ci.adoptium.net/job/Grinder/7978/tapTestReport/)

**Comment by sxa** [comment]
So the above suggests it's inconsistent but more reliable on the odroid systems. It may be worth attempting to run these in a docker container on the odroid system to see if it's in any way related to running in a container but I feel we need to get these running reliably on all systems.

Other dacapo related issues:
- https://github.com/adoptium/aqa-tests/issues/4276
- https://github.com/adoptium/aqa-tests/issues/4745

**Comment by sxa** [comment]
TC grinder 3872 on the containerized system gave problems:
```
11:48:37  TESTING:
11:48:37  Using scaled threading model. 16 processors detected, 16 threads used to drive the workload, in a possible range of [1,100]
11:48:37  ===== DaCapo 9.12-MR1 xalan starting =====
11:48:38  ; SystemID: [file:///home/jenkins/workspace/Grinder/aqa-tests/TKG/output_16989257163928/dacapo-xalan_0/./scratch/xalan/xmlspec.xsl;](file:///home/jenkins/workspace/Grinder/aqa-tests/TKG/output_16989257163928/dacapo-xalan_0/scratch/xalan/xmlspec.xsl;) Line#: 1440; Column#: 35
11:48:38  javax.xml.transform.TransformerException: java.lang.NullPointerException
11:48:38  	at org.apache.xalan.templates.ElemValueOf.execute(ElemValueOf.java:293)
```
which was ultimately an `ArrayIndexOutOfBounds` exception

Failed 2/10 times on the mythic machine in TC so there is definitely some flakiness here.

**Comment by andrew-m-leonard** [comment]
Raised upstream bug with analysis : https://bugs.openjdk.org/browse/JDK-8391660

**Comment by andrew-m-leonard** [comment]
Deep dive with IBM Bob:

# Upstream Bug Reports

Two related issues identified during DaCapo xalan benchmark failure analysis on Ubuntu 24.04 / aarch32 Linux.

---

## Bug 1 — OpenJDK aarch32 Port

**File upstream:** [bugs.openjdk.org](https://bugs.openjdk.org) — component `hotspot`, subcomponent `runtime`

**Title:** DaCapo xalan benchmark fails with "Can not setRoot on a non-iterated NodeSequence" on Linux/aarch32

**Component:** hotspot / runtime / aarch32 port  
**File:** `hotspot/src/os_cpu/linux_aarch32/vm/orderAccess_linux_aarch32.inline.hpp`

### Description

The `arm_atomic_load` and `arm_atomic_store` macros in the Linux/aarch32 `OrderAccess` implementation use plain C loads/stores with a *separate* memory barrier:

```cpp
#define arm_atomic_load(S, D, X) { \
    *(D) = *(S); if (X == ARM_ATOMIC_ACQUIRE) READ_MEM_BARRIER; \
}
#define arm_atomic_store(D, S, X) { \
    if (X == ARM_ATOMIC_RELEASE) WRITE_MEM_BARRIER; *(D) = *(S); \
}
```

On ARMv7, the memory access and the barrier are separate instructions. This leaves a window between the plain load/store and the barrier where another CPU core can observe a stale value, violating the Java Memory Model's happens-before guarantee that `synchronized` block entry/exit must enforce.

The aarch64 implementation correctly uses GCC `__atomic_load`/`__atomic_store` builtins with `__ATOMIC_ACQUIRE`/`__ATOMIC_RELEASE`, which map to barrier-coupled instructions. The aarch32 implementation should match this.

### Observed Failure

Multi-threaded Java programs that rely on `synchronized` for inter-thread visibility can observe races that do not occur on aarch64 or x86. Observed with the DaCapo `xalan` benchmark on Ubuntu 24.04/aarch32 with 4 CPUs, producing the following failure:

```
Using scaled threading model. 4 processors detected, 4 threads used to drive the workload
===== DaCapo 23.11-MR2-chopin-minimal xalan starting =====
javax.xml.transform.TransformerException: java.lang.RuntimeException: Programmer assertion is incorrect! - Can not setRoot on a non-iterated NodeSequence!
	at org.apache.xalan.transformer.TransformerImpl.executeChildTemplates(TransformerImpl.java:2407)
	at org.apache.xalan.templates.ElemLiteralResult.execute(ElemLiteralResult.java:1376)
	at org.apache.xalan.templates.ElemApplyTemplates.transformSelectedNodes(ElemApplyTemplates.java:395)
	at org.apache.xalan.templates.ElemApplyTemplates.execute(ElemApplyTemplates.java:178)
	at org.apache.xalan.transformer.TransformerImpl.executeChildTemplates(TransformerImpl.java:2402)
	at org.dacapo.xalan.XSLTBench$XalanWorker.run(XSLTBench.java:157)
Caused by: java.lang.RuntimeException: Programmer assertion is incorrect! - Can not setRoot on a non-iterated NodeSequence!
	at org.apache.xpath.Expression.assertion(Expression.java:423)
	at org.apache.xpath.axes.NodeSequence.setRoot(NodeSequence.java:275)
	at org.apache.xpath.axes.LocPathIterator.asIterator(LocPathIterator.java:269)
	at org.apache.xalan.templates.ElemApplyTemplates.transformSelectedNodes(ElemApplyTemplates.java:207)
	at org.apache.xalan.templates.ElemApplyTemplates.execute(ElemApplyTemplates.java:178)
	at org.apache.xalan.transformer.TransformerImpl.executeChildTemplates(TransformerImpl.java:2402)
	... 26 more
```

The same test passes consistently on Ubuntu 20.04/aarch32 with 8 CPUs and on aarch64 with the same JDK version. The failure is specific to aarch32 Linux with 4 or fewer CPUs providing true parallelism.

### Fix

Replace the hand-rolled macros with GCC `__atomic` builtins (requires GCC 4.7+), matching the aarch64 implementation:

```cpp
#define ARM_ATOMIC_ACQUIRE __ATOMIC_ACQUIRE
#define ARM_ATOMIC_RELEASE __ATOMIC_RELEASE
#define ARM_ATOMIC_RELAXED __ATOMIC_RELAXED

// Valid for values up to pointer size. DO NOT USE for jlong on 32-bit!
#define arm_atomic_load(S, D, X) \
    __atomic_load(S, D, X)
#define arm_atomic_store(D, S, X) \
    __atomic_store(D, S, X)
```

**Patch:** `fix-orderaccess-aarch32.patch`

---

## Bug 2 — Apache Xalan-J

**File upstream:** [issues.apache.org/jira](https://issues.apache.org/jira/projects/XALANJ) — project `XALANJ`

**Title:** Concurrent XSLT transformation throws "Can not setRoot on a non-iterated NodeSequence" RuntimeException

**Component:** `org.apache.xpath.axes.LocPathIterator`  
**Affects:** Xalan-J 2.7.2, 2.7.3

### Description

`LocPathIterator` maintains an `IteratorPool` (`m_clones`) of reusable iterator instances on the compiled stylesheet expression, shared across all `Transformer` instances created from the same `Templates`. Both `execute()` and `asIterator()` check out an instance from the pool and wrap it in an `XNodeSet`:

```java
XNodeSet iter = new XNodeSet((LocPathIterator)m_clones.getInstance());
iter.setRoot(contextNode, xctxt);
return iter;
```

`XNodeSet(DTMIterator)` stores the passed iterator directly as `m_iter` — a shallow reference, not a clone. When the `XNodeSet` is consumed and `detach()` is called, `LocPathIterator.detach()` returns `this` to the pool via `m_clones.freeInstance(this)`. However, the `XNodeSet` that wraps it may still be alive on another thread with `m_iter` pointing to the now-pooled instance.

If a second thread concurrently calls `getInstance()` and receives the same iterator, it wraps it in a new `XNodeSet` and calls `setRoot()` on the same instance the first thread's `XNodeSet` still holds. This is a data race on a shared mutable object.

Note: `executeCharsToContentHandler()` and `asNode()` are not affected — they call `clone.detach()` themselves before returning, so the instance is fully done before it re-enters the pool.

### Observed Failure

Consistent crash under multi-threaded XSLT transformation (DaCapo `xalan` benchmark, 4 threads, Ubuntu 24.04/aarch32). The race is timing-sensitive and may be masked on fewer-core hardware or on architectures where `synchronized` barriers accidentally serialise access.

```
Caused by: java.lang.RuntimeException: Programmer assertion is incorrect! - Can not setRoot on a non-iterated NodeSequence!
	at org.apache.xpath.Expression.assertion(Expression.java:423)
	at org.apache.xpath.axes.NodeSequence.setRoot(NodeSequence.java:275)
	at org.apache.xpath.axes.LocPathIterator.asIterator(LocPathIterator.java:269)
	at org.apache.xalan.templates.ElemApplyTemplates.transformSelectedNodes(ElemApplyTemplates.java:207)
	at org.apache.xalan.templates.ElemApplyTemplates.execute(ElemApplyTemplates.java:178)
	at org.apache.xalan.transformer.TransformerImpl.executeChildTemplates(TransformerImpl.java:2402)
	at org.apache.xalan.transformer.TransformerImpl.applyTemplateToNode(TransformerImpl.java:2272)
	at org.apache.xalan.transformer.TransformerImpl.transformNode(TransformerImpl.java:1358)
	at org.apache.xalan.transformer.TransformerImpl.transform(TransformerImpl.java:711)
	at org.apache.xalan.transformer.TransformerImpl.transform(TransformerImpl.java:1275)
	at org.apache.xalan.transformer.TransformerImpl.transform(TransformerImpl.java:1253)
	at org.dacapo.xalan.XSLTBench$XalanWorker.run(XSLTBench.java:157)
```

### Fix

Call `allowDetachToRelease(false)` on the checked-out iterator before handing ownership to `XNodeSet`, so that `detach()` discards rather than pools the instance. This has no functional impact on XPath evaluation results — pool reuse continues to work for `asNode()` and `cloneWithReset()` which manage their own iterator lifecycle.

```java
// execute()
LocPathIterator clone = (LocPathIterator)m_clones.getInstance();
clone.allowDetachToRelease(false);
XNodeSet iter = new XNodeSet(clone);
iter.setRoot(xctxt.getCurrentNode(), xctxt);
return iter;

// asIterator()
LocPathIterator clone = (LocPathIterator)m_clones.getInstance();
clone.allowDetachToRelease(false);
XNodeSet iter = new XNodeSet(clone);
iter.setRoot(contextNode, xctxt);
return iter;
```

**Patch:** `fix-xalan-locpathiterator.patch`

**Comment by andrew-m-leonard** [comment]
Xalan-J Jira bug opened: https://issues.apache.org/jira/browse/XALANJ-2850
